drivers/ksz8851snl/ksz8851_netdev.c

```diff
@@ -0,0 +1,622 @@
+
+		/* Send Read fifo command */
+		spi_transfer_bytes(dev->spi, tx_cmd, NULL, 1);
+
+		/* Dummy eight byte read to avoid dummy bytes - status word - length word */
+		spi_transfer_bytes(dev->spi, NULL, NULL, 8);
+
+		/* 2 bytes to align to word */
+		spi_transfer_bytes(dev->spi, NULL, NULL, 2);
+
+		/* Transfer Data */
+		spi_transfer_bytes(dev->spi, NULL, buf, (len - 6));
+
+		/* Maintain alignment to 4-byte boundaries */
+		spi_transfer_bytes(dev->spi, NULL, NULL, len % 4);
+
+		/* Release SNL and SPI control */
+		gpio_set(dev->cs);
+		spi_release(dev->spi);
+
+		/* disable access to the Rx queue */
+		reg_bits_clear(dev, KSZ8851_RXQCR_REG, RXQCR_QUEUE_ACCESS_EN);
+	}
+	unlock(dev);
+	return frame_size;
+}
+
+/* This is IRQ Handler for KSZ8851SNL privileged IRQ keep short */
+static void ksz8851snl_pIsr(void *netdev)
+{
+	ksz8851snl_t *dev = (ksz8851snl_t*)netdev;
+
+	/* disable interrupt line TODO: Check if I need both or either the MCU or KSZ will do */
+	gpio_irq_disable(dev->int_pin);
+	reg_set(dev, KSZ8851_IER_REG, 0x0000);
+
+	/* call netdev2 even to init frame receive process */
+	dev->netdev.event_callback((netdev2_t*)dev, NETDEV2_EVENT_ISR, NULL);
+}
+
+static void ksz8851snl_isr(netdev2_t *netdev)
+{
+	ksz8851snl_t *dev = (ksz8851snl_t*)netdev;
+	uint16_t isr_mask;
+
+	lock(dev);
+
+	/* Read interrupt status register */
+	isr_mask = reg_get(dev, KSZ8851_ISR_REG);//TODO: DO I need to do this before I reset the isr
+
+	/* Interrupt Handlers */
+
+	/*  IRQ */
+	if (isr_mask & ISR_LDIS) {
+		uint16_t temp_irq = reg_get(dev, KSZ8851_PMECR_REG);
+		temp_irq |= (1 << 3);
+		temp_irq &= (~(0x000D << 2));
+		reg_set(dev, KSZ8851_PMECR_REG, temp_irq);
+	}
+
+	/* Link Change IRQ */
+	if (isr_mask & ISR_LCIS) {
+		/* Check and report what the current status of the link */
+		uint16_t link = reg_get(dev, KSZ8851_P1SR_REG);
+
+		netdev2_event_t event = (link & P1SR_LINK_GOOD) ?
+					NETDEV2_EVENT_LINK_UP :
+					NETDEV2_EVENT_LINK_DOWN;
+		if (link & P1SR_LINK_GOOD)
+			DEBUG(SECT_NAME "Link Status is Up\n");
+		else
+			DEBUG(SECT_NAME "Link Status is Down\n");
+
+		reg_set(dev, KSZ8851_ISR_REG, ISR_LCIS);           //TODO I need to look into unlocking here
+		dev->netdev.event_callback(netdev, event, NULL);        //could be wrong first variable
+	}
+	/* Packet Transmited IRQ */
+	if (isr_mask & ISR_TXIS) { /* TODO: Get rid of the Tx Irq */
+		reg_set(dev, KSZ8851_ISR_REG, ISR_TXIS);
+	}
+	/* Packet Received IRQ */
+	if (isr_mask & ISR_RXIS) {
+		/* Reset the IRQ Flag */
+		reg_set(dev, KSZ8851_ISR_REG, ISR_RXIS);
+		uint16_t frames_avail = (reg_get(dev, KSZ8851_RXFCTR_REG) >> 8);
+
+		while (frames_avail) {
+			unlock(dev);
+			netdev->event_callback(netdev, NETDEV2_EVENT_RX_COMPLETE, NULL);
+			lock(dev);
+			frames_avail--;
+		}
+	}
+	/* Restore IRQs and Release Mutex */
+	gpio_irq_enable(dev->int_pin);
+	reg_set(dev, KSZ8851_IER_REG, IER_DEFAULT_CONFIG);
+	unlock(dev);
+}
+
+static int ksz8851snl_reset(ksz8851snl_t *dev)
+{
```

Comment:
is there any benefit of a soft reset over a hard reset? If not, I would just do the latter to save some bytes of ROM...